<h2>FIN5622 Project #1
    <p class="text-primary">Loan Default Prediction with Logistic Regression</p>
</h2>

<div class = "alert alert-block alert-danger">
    <b>Due Date</b>: March 19th, 2021, 11:59PM EST
</div>  

<div class = "panel panel-success">
    <div class = "panel-heading"> 
    <span style="color:black"><b>Project Introduction</b>
    </div>
    <div class = "panel-body"> 
    In this project we will build a logistic regression model to predict borrower defaults on a set of historical loan performance dataset.
    </div>
</div>

<div class = "panel panel-info">
    <div class = "panel-heading"> 
    <span style="color:black"><b>Instructions</b>
    </div>
    <div class = "panel-body"> 
    Please code you solutions directly in this Notebook under each task in the cells provided below. Tasks are intended to be done in order with results from previous tasks informing work in later tasks. Include all code used to produce your results, and support your answers with comments if neccessary. You may feel free to add additional Code and/or Markdown cells as needed. Code must be written using <b>Python 3</b>.
    <br><br>Please order the cells in your Notebook in the order in which they are intended to run! It is expected that your Notebook is able to run in sequence from start to finish without any errors (before submitting check by: <code>Kernel > Restart & Run All</code>).
    <br><br>Once ready, please submit your assignment as an <b>.ipynb file</b> (<code>File > Download as > Notebook</code>). You must name your files in the following format: <b><code>FIN5622_Project_1_Richard_Lee_tuc12345.ipynb</code></b>.
    </div>
</div>

---
<div class = "alert alert-block alert-warning">
    <b>Note</b>: Import all libraries that you plan on using in the code block below.
</div>

In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 1</b>
    </div>
    <div class = "panel-body"> 
    Read in the dataset "loan_performance_data_2016.csv" into a dataframe called <code>loans</code> using the pandas <code>read_csv</code> function. Check that there are 405,180 loans present in the dataframe.
    <br><br>
    Note that there are 16 feature columns and 1 target column (i.e. "loan_default") in this dataset. Please review the file "Loan_Data_Dictionary.xlsx" for definitions on each column. Pick a variable that you believe will be an important explanatory variable for loan defaults and explain why.
    <p style="text-align:right;">[5pts]</p>
    </div>
</div>

In [23]:
path = '../Data'

In [24]:
loans = pd.read_csv(f'{path}/loan_performance_data_2016.csv')
loans

loan_amnt  loan_term  int_rate  fico    dti  annual_income  \
0           7.200        3.0      5.32   750  19.05           49.0   
1          12.000        5.0     11.99   670   6.59           49.0   
2          11.425        3.0     19.53   730  35.56           26.0   
3          17.500        3.0     13.67   665  38.29           46.0   
4          22.875        5.0     17.27   670  11.75           65.0   
...           ...        ...       ...   ...    ...            ...   
405175     16.000        5.0     12.79   665  12.25          150.0   
405176     24.000        5.0     12.79   665  19.61           95.0   
405177     40.000        5.0     10.49   705  12.75          227.0   
405178     24.000        5.0     14.49   660  18.30          110.0   
405179     14.000        5.0     14.49   660  23.36           95.0   

        income_verification  home_mortgage  home_own  home_rent  \
0                         0              1         0          0   
1                         0              1         0          0   
2                         0              0         0          1   
3                         0              0         0          1   
4                         0              0         0          1   
...                     ...            ...       ...        ...   
405175                    0              0         0          1   
405176                    0              1         0          0   
405177                    1              1         0          0   
405178                    0              0         0          1   
405179                    1              1         0          0   

        debt_consolidation  coborrower  num_accounts  revol_bal  revol_util  \
0                        1           0            11      9.309        18.4   
1                        0           0            10     12.152        50.6   
2                        0           0            11      2.096         7.2   
3                        1           0            33     18.847        62.6   
4                        1           0             6     25.200        90.0   
...                    ...         ...           ...        ...         ...   
405175                   0           0            12      7.700        55.0   
405176                   0           0             5     49.431        84.4   
405177                   0           0             5      8.633        64.9   
405178                   1           0            10     17.641        68.1   
405179                   0           0             8      7.662        54.0   

        derogatory_records  loan_default  
0                        0             0  
1                        0             1  
2                        0             0  
3                        0             0  
4                        0             0  
...                    ...           ...  
405175                   1             0  
405176                   0             0  
405177                   0             0  
405178                   1             1  
405179                   0             0  

[405180 rows x 17 columns]

In [25]:
len(loans)

405180

I think that an explanatory variable of whether or not a loan deafults is going to be the term of the loan. When a loan is longer term, there are more variables that can cause someone to default such as a crash of the market or the loan holder losing their job.

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 2</b>
    </div>
    <div class = "panel-body"> 
    We will be creating a logistic regression model to predict the "loan_default" column, with the rest of the columns as feature variables. 
    <br><br>
    Create a new variable called <code>loan_features</code> to store just the columns that are input variables, and another variable called <code>loan_target</code> to store the prediction target. Each array shoud have the following shapes: 
    <ul>
        <li><code>loan_features.shape = (405180, 16)</code></li>
        <li><code>loan_target.shape = (405180,)</code></li>
    </ul>
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [26]:
loan_features = loans.drop(['loan_default'], axis = 1).copy()
loan_features

#created variable to store all input variables in order to predict whether or not a loan will default

loan_amnt  loan_term  int_rate  fico    dti  annual_income  \
0           7.200        3.0      5.32   750  19.05           49.0   
1          12.000        5.0     11.99   670   6.59           49.0   
2          11.425        3.0     19.53   730  35.56           26.0   
3          17.500        3.0     13.67   665  38.29           46.0   
4          22.875        5.0     17.27   670  11.75           65.0   
...           ...        ...       ...   ...    ...            ...   
405175     16.000        5.0     12.79   665  12.25          150.0   
405176     24.000        5.0     12.79   665  19.61           95.0   
405177     40.000        5.0     10.49   705  12.75          227.0   
405178     24.000        5.0     14.49   660  18.30          110.0   
405179     14.000        5.0     14.49   660  23.36           95.0   

        income_verification  home_mortgage  home_own  home_rent  \
0                         0              1         0          0   
1                         0              1         0          0   
2                         0              0         0          1   
3                         0              0         0          1   
4                         0              0         0          1   
...                     ...            ...       ...        ...   
405175                    0              0         0          1   
405176                    0              1         0          0   
405177                    1              1         0          0   
405178                    0              0         0          1   
405179                    1              1         0          0   

        debt_consolidation  coborrower  num_accounts  revol_bal  revol_util  \
0                        1           0            11      9.309        18.4   
1                        0           0            10     12.152        50.6   
2                        0           0            11      2.096         7.2   
3                        1           0            33     18.847        62.6   
4                        1           0             6     25.200        90.0   
...                    ...         ...           ...        ...         ...   
405175                   0           0            12      7.700        55.0   
405176                   0           0             5     49.431        84.4   
405177                   0           0             5      8.633        64.9   
405178                   1           0            10     17.641        68.1   
405179                   0           0             8      7.662        54.0   

        derogatory_records  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0  
...                    ...  
405175                   1  
405176                   0  
405177                   0  
405178                   1  
405179                   0  

[405180 rows x 16 columns]

In [27]:
loan_target = loans.loc[:, 'loan_default'].copy()
loan_target

#created variable to store output data (ie. whether or not the loan defaulted)

0         0
1         1
2         0
3         0
4         0
         ..
405175    0
405176    0
405177    0
405178    1
405179    0
Name: loan_default, Length: 405180, dtype: int64

In [28]:
loan_features.shape

(405180, 16)

In [29]:
loan_target.shape

(405180,)

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 3</b>
    </div>
    <div class = "panel-body"> 
    Fit a logistic regression model using the data stored in <code>loan_features</code> and <code>loan_target</code>. 
    <br><br>
    Recall that this is done by calling <code>LogisticRegression()</code> from the <code>sklearn.linear\_model</code> module, and using the <code>fit()</code> method. Store the fitted model in a variable called <code>logistic_model</code>.
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [30]:
logistic_model = LogisticRegression(max_iter = 1000).fit(X = loan_features, y = loan_target)


<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 4</b>
    </div>
    <div class = "panel-body"> 
    Examine the model's fitted coefficients (stored in the <code>coef_</code> attribute), and answer the following questions:
    <ul>
        <li>How should we interpret the output of the logistic regression model?</li>
        <li>How should we interpret a positive coefficient in the logistic regression model? How about a negative coefficient?</li>
        <li>Are all the signs of the fitted coeffiecients in your logistic regression model for each variable as expected? If not, list the variables with unexpected signs and explain what the expected sign on the coefficient should be and why.</li>
    </ul>
    <p style="text-align:right;">[20pts]</p>
    </div>
</div>

In [31]:
pd.Series(logistic_model.coef_[0], index = loan_features.columns)

loan_amnt              0.011500
loan_term              0.299658
int_rate               0.092084
fico                  -0.005852
dti                    0.009946
annual_income         -0.001236
income_verification    0.103053
home_mortgage         -0.173258
home_own               0.013054
home_rent              0.172240
debt_consolidation    -0.083727
coborrower            -0.051549
num_accounts           0.012690
revol_bal             -0.004194
revol_util            -0.000674
derogatory_records     0.099752
dtype: float64

In this case the biggest factors are:
1: loan term (.29), 
2: home mortgage (-.17), 
3: home rent (.17), 
4: income verifcation (.1), 
5: interest rate(.09)

_____________________________________________________________________________

The data outputted from using the coef_ attribute are the betas for each of data feature. A positive beta means that as the value of the feature observation point increases, the target observation (whether or not the loan defaults) is more likely to represent a value of 1 which in this case mean the loan defaults. 

_________________________________________

[income verification] I would expect that if an observation had verified income that it would make them less likely to default vs more likely. Or at least be close to zero meaning having little impact.

[revolving balance & revolving utility]
I would expect a higher value for both of these to make you more likely to default vs less likely. I would expect that having a higher balance of money you owe would make you more likely to default on your loans than if your balance was lower and I would expect the same for your utility.
I understand that both of these values are close to zero and are of little impact, but with the amount of trials I would expect them to converge to their mean (which I assumed was positive).

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 5</b>
    </div>
    <div class = "panel-body"> 
    Calculate the logistic regression model's classification error on our dataset <code>loan_features</code> and <code>loan_target</code>. Recall that this is done by calling the <code>score()</code> method from <code>logistic_model</code>.
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [32]:
score = logistic_model.score(X = loan_features,y = loan_target)
score

0.8063749444691248

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 6</b>
    </div>
    <div class = "panel-body"> 
    Use the fitted logistic regression model saved in <code>logistic_model</code> to predict the loan default probabilities for the loans in <code>loan_features</code>. Recall that this is done by calling the <code>predict_proba()</code> method from logistic_model. 
    <br><br>
    Note that while the output of <code>predict_proba()</code> is a 2-d array, we only need the 2nd column, which is the probability of loan defaults, and should be a vector of numbers between 0 and 1. Index the 2nd columns and store the probabilities in a variable called <code>model_prob</code>.
    <br><br>
    Calculate the average predicted probability of default for the loans in <code>loan_features</code>. How close is it to the actual observed default probability for the loans in the test dataset <code>loan_target</code>?
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [33]:
model_prob = logistic_model.predict_proba(loan_features)[:, 1]
model_prob

array([0.04995092, 0.21498883, 0.30753727, ..., 0.19308043, 0.37739351,
       0.31448359])

In [34]:
len(model_prob)

405180

In [35]:
ld_prediction_mean = np.mean(model_prob)
ld_prediction_mean

0.19597987879686785

In [36]:
ld_actual_mean = np.mean(loan_target)
ld_actual_mean

0.19609803050496075

In [37]:
abs(ld_prediction_mean - ld_actual_mean)

0.00011815170809290065

In [38]:
RMSE = np.mean((model_prob - loan_target)**2)**.5
RMSE

0.37811593952355627

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 7</b>
    </div>
    <div class = "panel-body"> 
    Calculate the loan default predictions from the logistic regression model on <code>loan_features</code> by using the <code>predict()</code> method. Note that the results should be a vector of 1's and 0's. Store the predictions in a variable called <code>model_pred</code>. 
    <p style="text-align:right;">[10pts]</p>
    </div>
</div>

In [39]:
model_pred = logistic_model.predict(loan_features)
model_pred

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [40]:
model_pred == loan_target

0          True
1         False
2          True
3          True
4          True
          ...  
405175     True
405176     True
405177     True
405178    False
405179     True
Name: loan_default, Length: 405180, dtype: bool

In [41]:
np.mean(model_pred == loan_target)

0.8063749444691248

<div class = "panel panel-primary">
    <div class = "panel-heading"> 
    <span style="color:white"><b>Task 8</b>
    </div>
    <div class = "panel-body"> 
    Create the confusion matrix for the predictions in <code>model_pred</code> compared to the actual outcomes in <code>loan_target</code>. Recall that this can be done using the function <code>confusion_matrix()</code> from the <code>sklearn.metrics</code> module.
    <br><br>
    Interpret and explain the results of each of the 4 cells of the confusion matrix in the context of this project. 
    <p style="text-align:right;">[15pts]</p>
    </div>
</div>

In [42]:
pd.DataFrame(confusion_matrix(y_true = loan_target, y_pred = model_pred) / len(loan_target))

0         1
0  0.786426  0.017476
1  0.176149  0.019949

78.64% of the data are [non-defaults] and we predicted them to be [non-defaults]

1.74% of the data are [defaults] and we predicted them to be [non-defaults]

17.61% of the data are [non-defaults] and we predicted them to be [defaults]

1.99% of the data are [defaults] and we predicted them to be [defaults]